crates/pgt_workspace/src/workspace/server/parser.rs

```diff
@@ -0,0 +1,374 @@
+    /// Hence, we just clear the AST for the old statements and lazily load them when requested.
+    ///
+    /// * `params`: ChangeFileParams - The parameters for the change to be applied.
+    pub fn apply_change(&mut self, params: ChangeFileParams) {
+        for c in &self.doc.apply_file_change(&params) {
+            match c {
+                StatementChange::Added(added) => {
+                    tracing::debug!(
+                        "Adding statement: id:{:?}, text:{:?}",
+                        added.stmt,
+                        added.text
+                    );
+                    self.cst_db.add_statement(&added.stmt, &added.text);
+                }
+                StatementChange::Deleted(s) => {
+                    tracing::debug!("Deleting statement: id {:?}", s,);
+                    self.cst_db.remove_statement(s);
+                    self.ast_db.clear_statement(s);
+                    self.sql_fn_db.clear_statement(s);
+                }
+                StatementChange::Modified(s) => {
+                    tracing::debug!(
+                        "Modifying statement with id {:?} (new id {:?}). Range {:?}, Changed from '{:?}' to '{:?}', changed text: {:?}",
+                        s.old_stmt,
+                        s.new_stmt,
+                        s.change_range,
+                        s.old_stmt_text,
+                        s.new_stmt_text,
+                        s.change_text
+                    );
+
+                    self.cst_db.modify_statement(s);
+                    self.ast_db.clear_statement(&s.old_stmt);
+                    self.sql_fn_db.clear_statement(&s.old_stmt);
+                }
+            }
+        }
+    }
+
+    pub fn get_document_content(&self) -> &str {
+        &self.doc.content
+    }
+
+    pub fn document_diagnostics(&self) -> &Vec<SDiagnostic> {
+        &self.doc.diagnostics
+    }
+
+    pub fn find<'a, M>(&'a self, id: StatementId, mapper: M) -> Option<M::Output>
+    where
+        M: StatementMapper<'a>,
+    {
+        self.iter_with_filter(mapper, IdFilter::new(id)).next()
+    }
+
+    pub fn iter<'a, M>(&'a self, mapper: M) -> ParseIterator<'a, M, DefaultFilter>
+    where
+        M: StatementMapper<'a>,
+    {
+        self.iter_with_filter(mapper, DefaultFilter)
+    }
+
+    pub fn iter_with_filter<'a, M, F>(&'a self, mapper: M, filter: F) -> ParseIterator<'a, M, F>
+    where
+        M: StatementMapper<'a>,
+        F: StatementFilter<'a>,
+    {
+        ParseIterator::new(self, mapper, filter)
+    }
+
+    #[allow(dead_code)]
+    pub fn count(&self) -> usize {
+        self.iter(DefaultMapper).count()
+    }
+}
+
+pub trait StatementMapper<'a> {
+    type Output;
+
+    fn map(
+        &self,
+        parser: &'a Parser,
+        id: StatementId,
+        range: TextRange,
+        content: &str,
+    ) -> Self::Output;
+}
+
+pub trait StatementFilter<'a> {
+    fn apply(&self, id: &StatementId, range: &TextRange) -> bool;
```

Comment:
Nit: If I were to implement a filter, I would have to dive into the source code of `ParseIterator` to find out whether this should return `true` or `false` when I want the statement included.
How about `filter_in` or `filter_out` or `predicate` ?